### PR TITLE
check whether specified launcher & launcher exec are available + fix faulty tests

### PR DESCRIPTION
**Pull request by boegel**
fix for #136

@stdweird I would like to have this solved before I look into SLURM support.

### CODE REVIEW

File: lib/vsc/mympirun/mpi/mpi.py [hunk 1]
```diff
@@ -927,8 +931,17 @@ def make_mpiexec_hydra_options(self):
             else:
                 launcher_exec = self.get_rsh()
 
-            self.mpiexec_options.append("-%s-exec %s" % (self.HYDRA_LAUNCHER_NAME, launcher_exec))
-
+            launcher_exec_opt = '-%s-exec %s' % (self.HYDRA_LAUNCHER_NAME, launcher_exec)
+            if which(launcher_exec):
+                # only specify launcher exec if specified command is available;
+                # only use pbsssh as launcher exec if pbsdsh is also available
+                if launcher_exec == 'pbsssh' and which('pbsdsh') is None:
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: this is logic for `pbs.rm`

File: lib/vsc/mympirun/rm/pbs.py [hunk 2]
```diff
@@ -37,13 +38,22 @@ class PBS(Sched):
...
-    HYDRA_LAUNCHER_EXEC = 'pbsssh'
     HYDRA_RMK = ['pbs']
 
+    def __init__(self, *args, **kwargs):
+        """PBS constructor"""
+        super(PBS, self).__init__(*args, **kwargs)
+
+        # pbsssh is only used if native support for pbsdsh is not supported in the MPI library;
+        # e.g. Intel MPI v5.x and newer supports using pbsdsh as launcher natively, no need for pbsssh wrapper
+        if which('pbsssh') and which('pbsdsh'):
```
Comment (stdweird) [inline lib/vsc/mympirun/rm/pbs.py]: use a constant or the same string

File: lib/vsc/mympirun/rm/pbs.py [hunk 3]
```diff
@@ -37,13 +38,22 @@ class PBS(Sched):
...
+            self.log.debug("Both 'pbsssh' and 'pbsdsh' found, so use 'pbsssh' as remote shell command.")
+            self.RSH_LARGE_CMD = 'pbsssh'
+            self.RSH_LARGE_LIMIT = 'pbsssh'
+            self.HYDRA_LAUNCHER_EXEC = 'pbsssh'
+        else:
+            self.log.debug("Either 'pbsssh' or 'pbsdsh' not available, not using 'pbsssh' as remote shell command.")
```
Comment (stdweird) [inline lib/vsc/mympirun/rm/pbs.py]: can you also add what is being used to the message?

File: test/end2end.py [hunk 4]
```diff
@@ -320,11 +328,10 @@ def test_env_variables(self):
             self.assertTrue(key in out, "%s is not in out" % key)
 
         regex = r'.*-envlist [^ ]*USER.*'
-        self.assertTrue(regex.find(out), "Variablesprefix USER isn't passed to mympirun script env")
```
Comment (stdweird) [inline test/end2end.py]: ha, that's a nice one. how did this one ever pass?

Comment (boegel) [reply test/end2end.py]:
```
>>> 'foo'.find('bar')
-1
>>> bool('foo'.find('bar'))
True
```

Comment (stdweird) [reply test/end2end.py]: `-1` should have guessed that.

**Pull request merged**